### ISSUE TITLE
Completion of module variables in module block

### ISSUE THREAD

**Issue by pecigonzalo** [hashicorp/terraform-ls#93]
I could not find any issues or references to it, but does autocompletion work from modules? We use a lot of them, and it does not seem to work for us.

**Comment by radeksimko** [comment]
Hi @pecigonzalo 
I'm aware that this may be a bit too late to ask given that you filed the issue in December 2018, but do you remember what the context was?

i.e. what do you mean by "completion from modules"?

There is a few aspects of modules where completion comes to mind:

 - opening module's root in IDE and getting completion inside configuration that belongs to the module
   - in theory this should already work out of the box, but it's probably less common for folks to `terraform init` inside the module directory and more importantly you'd have to open the folder explicitly, _not_ it's parent folder where the module is referenced, because the server needs the relevant providers to be downloaded, so it can obtain the schema (which in turn gives it the completion data) - see https://github.com/hashicorp/terraform-ls/issues/83 to read more about how we plan to address that
 - completing arguments within `module` block (module reference)
   - this doesn't work yet, and I think it generally falls under https://github.com/hashicorp/terraform-ls/issues/36
 - completing `source` of `module` block (module reference)
   - this doesn't work and I'm unsure what the ideal UX be here. We could probably implement this as Terraform Registry search for modules there, but we might need to also account for private module Registries which are firewalled-off and I'm not sure if we have a mechanism for discovering the existence of such registry, so we can search in it

**Comment by pecigonzalo** [comment]
Unfortunately no and I was not really good at describing it initially, ill give it a look tomorrow and try to repro.

**Comment by pecigonzalo** [comment]
@radeksimko I got around to remember what this was about. As we use tons of modules, we wanted to be able to autocomplete module outputs when typing them.

Lets say I have module for a vpc like
```hcl
resource "aws_vpc" "default" {
  [...]
}

output "vpc_id" {
  value       = "${aws_vpc.default.id}"
  description = "The VPC ID"
}
```

and some code like
```hcl
module "vpc" {
  [...]
}

output "id" {
  value = module.vpc.vpc_id
}
```


When typing `module.vpc.` I would like to get suggestions for the possible attributes of that module, in this example `vpc_id`.
Ideally when consuming the module and writing inside the module block, it would be great to also get suggestions for the attributes .

**Comment by radeksimko** [comment]
@pecigonzalo Understood - that's another valid case I forgot to mention above. 👍 

I think this would generally fall under https://github.com/hashicorp/terraform-ls/issues/38 which will come after https://github.com/hashicorp/terraform-ls/issues/36 as we need to teach the parser how to parse module outputs first.

**Comment by paultyng** [comment]
Probably dependent upon this upstream issue: https://github.com/hashicorp/terraform-config-inspect/pull/24

**Comment by lorengordon** [comment]
Does this request for autocompletion of module outputs also extend to attributes of those outputs? For when they are complex objects, such as entire resources? E.g.

```
resource aws_vpc this{
  [...]
}

output vpc {
  value       = aws_vpc.this
  description = "The VPC object"
}
```

**Comment by radeksimko** [comment]
FYI: We have just released [`v0.9.0`](https://github.com/hashicorp/terraform-ls/releases/tag/v0.9.0) which provides basic completion for the static attributes within `module` block.

Completion for variables is yet to come and outputs fall under a slightly bigger task of supporting expressions which is tracked under https://github.com/hashicorp/hcl-lang/issues/2

**Comment by mwarkentin** [comment]
If it's helpful (probably depends how long the follow up takes), one of my coworkers wrote a vscode extension which adds autocomplete for module variables and outputs: https://github.com/mgtrrz/terraform-completer

**Comment by radeksimko** [comment]
This was implemented as part of https://github.com/hashicorp/terraform-ls/pull/551 and released in `0.20.0`
Module output completion was also implemented in https://github.com/hashicorp/terraform-ls/pull/593 and shipped as part of the same release.

Further improvements are planned in this area and I'd recommend reviewing the list and/or subscribing to this meta issue if you are interested: https://github.com/hashicorp/vscode-terraform/issues/715

Pre-filling required (module) attributes is also tracked under https://github.com/hashicorp/vscode-terraform/issues/719
